### PR TITLE
systemd: support systemd.tmpfiles.settings

### PR DESCRIPTION
**Pull request by phanirithvij**
copied verbatim from nixpkgs

~since this is writing to a single file, I am adding comments for each settings section to distinguish it.~ edit: it is writing to different files, one per settings option.

cc: @r-vdp

#146

### CODE REVIEW

**Comment by r-vdp** [comment]
Thanks!

Did you copy this code over from nixpkgs or you wrote it from scratch?
In the long run, it would be nice if we could factor out this code and reuse it between here and nixpkgs. But that's a longterm goal.

I'm also wondering if we should actually generate multiple files, like we do in nixpkgs, which gives more control to users, and it makes sure that modules that rely on this, are portable. You think that that would be feasible?

**Comment by r-vdp** [comment]
A test would also be nice.

**Comment by phanirithvij** [comment]
Yes I copied it from nixpkgs, as I don't know how to reuse it.

I'll see if I can get it working with multiple files same as in nixpkgs. And it test as well.

**Comment by phanirithvij** [comment]
Let me preface this by saying I don't have much experience with rust.

In tmp_files.rs activate function

```rust
    let mut cmd = process::Command::new("systemd-tmpfiles");
    cmd.arg("--create")
        .arg("--remove")
        .arg("/etc/tmpfiles.d/00-system-manager.conf");
```

Here, I need to change it to something like `systemd-tmpfiles --create --remove /etc/tmpfiles.d/*.conf`
But the existing ubuntu (host os) already has some services and tmpfiles configured, so it affects the existing tmpfiles.
- So I read from `etc_tree` and got the `tmpfiles.d` entry and looped over the files inside it.

I heard `unwrap` should be avoided, because it panics at runtime.
In this case since `/etc/tmpfiles.d/00-system-manager.conf` is guaranteed to exist, meaning `/etc/tmpfiles.d` for sure will be in `etc_tree`, and thus we shouldn't panic at all?

File: nix/modules/tmpfiles.nix [hunk 1]
```diff
@@ -15,13 +148,60 @@ in
...
+        name = "tmpfiles.d";
+        paths = map (p: p + "/lib/tmpfiles.d") config.systemd.tmpfiles.packages;
+        postBuild = ''
+          for i in $(cat $pathsPath); do
+            (test -d "$i" && test $(ls "$i"/*.conf | wc -l) -ge 1) || (
+              echo "ERROR: The path '$i' from systemd.tmpfiles.packages contains no *.conf files."
+              exit 1
+            )
+          done
+        '';
```
Comment (phanirithvij) [inline nix/modules/tmpfiles.nix:187]: Here it differs from nixpkgs version

orig nixpkgs version https://github.com/NixOS/nixpkgs/blob/01ddc67fd61124e3c9df1438e6874731af6d3f06/nixos/modules/system/boot/systemd/tmpfiles.nix#L240-L243

```nix
        '' + concatMapStrings (name: optionalString (hasPrefix "tmpfiles.d/" name) ''
          rm -f $out/${removePrefix "tmpfiles.d/" name}
        '') config.system.build.etc.passthru.targets;
      }) + "/*";
```

- there is no support for globs in system-manager's `environment.etc.source` so `+ "/*";` is out.
- I don't know what config.system.build.etc.passthru.targets is tbh

Comment (phanirithvij) [reply nix/modules/tmpfiles.nix:187]:
```nix
  etc' = lib.filter (f: f.enable) (lib.attrValues config.environment.etc);

  etc = pkgs.runCommandLocal "etc" {
    # This is needed for the systemd module
    passthru.targets = map (x: x.target) etc';
  } /* sh */ ''
```

from `nixos/modules/system/etc/etc.nix`

I don't know if it is needed for system-manager?

File: src/activate/tmp_files.rs [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
+        .nested
+        .get("etc")
+        .unwrap()
+        .nested
+        .get("tmpfiles.d")
+        .unwrap()
+        .nested
+        .iter()
+        .map(|(_, node)| node.path.to_string_lossy().to_string())
+        .collect::<Vec<_>>();
```
Comment (r-vdp) [inline src/activate/tmp_files.rs:20]:
```suggestion
    let conf_files = etc_tree
        .nested
        .get("etc")
        .and_then(|etc| etc.nested.get("tmpfiles.d"))
        .map_or(vec![], |tmpfiles_d| {
            tmpfiles_d
                .nested
                .iter()
                .map(|(_, node)| node.path.to_string_lossy().to_string())
                .collect::<Vec<_>>()
        });
```

**Review by r-vdp — CHANGES_REQUESTED** [review]
One small change to avoid having to call `unwrap`, besides that, this looks good to me!

It's really unfortunate that we need to copy all this code from nixpkgs, but I don't think we have a real alternative currently. If at some point this project gets more traction, we could have a discussion in the community to see if we can have some sort of re-usable modules that we can share to lower the maintenance burden.

**Pull request merged**